### PR TITLE
drivers: tla2021: Add driver

### PR DESCRIPTION
**Pull request by casparfriedrich**
This adds a driver for Texas Instruments Cost-Optimized, Ultra-Small, 12-Bit, System-Monitoring ADCs. Currently only TLA2021 is supported, TLA2022 and TLA2024 may follow based on this driver.

### CODE REVIEW

File: drivers/adc/adc_tla2021.c
Comment (cfriedt) [inline drivers/adc/adc_tla2021.c:92]: Might be better to use a spin lock here in case the call is made from isr context

File: drivers/adc/adc_tla2021.c
Comment (cfriedt) [inline drivers/adc/adc_tla2021.c:237]: I have not seen this API used in any driver context. For that reason, I would guess that it's probably not the best choice.

@anangl - can you provide some advice?

File: drivers/adc/adc_tla2021.c
Comment (dp7hgh7) [reply drivers/adc/adc_tla2021.c:92]: This driver motivates i2c... Is it even possible to call this from isr? We might need to return an error here.

File: drivers/adc/adc_tla2021.c
Comment (dp7hgh7) [reply drivers/adc/adc_tla2021.c:237]: I think events are pretty new. Do you see a different way to get a somewhat stable sample rate?

File: drivers/adc/adc_tla2021.c
Comment (dp7hgh7) [reply drivers/adc/adc_tla2021.c:92]: Now it returns `-EWOULDBLOCK` if the read function was called from within an isr.

File: drivers/adc/adc_tla2021.c
Comment (cfriedt) [reply drivers/adc/adc_tla2021.c:239]: @dp7hgh7 - a lot of the sensor drivers use a thread or workqueue I would check those out. There might even be a doc about it.

I just thought I would ask the ADC maintainer (@anangl) first to see if there was a preferred way of doing things for ADC drivers.

Comment (dp7hgh7) [reply drivers/adc/adc_tla2021.c:239]: I see. I'll have a look, thanks!

Comment (dp7hgh7) [reply drivers/adc/adc_tla2021.c]: I reworked my solution. Actually I completely overlooked the `adc_context.h`... It's now used and very close the the adsxxxx drivers already in the tree.

File: drivers/adc/adc_tla2021.c
Comment (cfriedt) [inline drivers/adc/adc_tla2021.c:14]: The preferred order is to put iso standard headers above zephyr headers, but not a blocking issue.

File: drivers/adc/adc_tla2021.c
Comment (cfriedt) [inline drivers/adc/adc_tla2021.c:177]: Nice👌

File: drivers/adc/adc_tla2021.c
Comment (cfriedt) [inline drivers/adc/adc_tla2021.c:46]: I would suggest adding these to Kconfig, but also I defer to @anangl in case there is already some established convention here.

File: drivers/adc/adc_tla2021.c
Comment (cfriedt) [inline drivers/adc/adc_tla2021.c:330]: Was just looking for this - good work 👍

**Review by cfriedt — CHANGES_REQUESTED** [review]
Looks good to me aside from 1 minor request - please add thread parameters to Kconfig.

File: drivers/adc/adc_tla2021.c
Comment (fabiobaltieri) [inline drivers/adc/adc_tla2021.c:74]: nit: just just `value`, not sure what convention is the trailing `_` trying to follow but I'm not aware of any here.

File: drivers/adc/adc_tla2021.c
Comment (fabiobaltieri) [inline drivers/adc/adc_tla2021.c:69]: `tla2021_read`? I'd suggest keeping the prefix even if the function is static, makes it easier to spot the specific one in a backtrace or to add a breakpoint.

File: drivers/adc/adc_tla2021.c
Comment (fabiobaltieri) [inline drivers/adc/adc_tla2021.c:116]: Personal preference but, you may as well just use the underlying function directly, finding these do not add anything to the code readability. Up to you whether to change them.

File: drivers/adc/adc_tla2021.c
Comment (fabiobaltieri) [inline drivers/adc/adc_tla2021.c:145]: Add a #define for 12?

File: drivers/adc/adc_tla2021.c
Comment (fabiobaltieri) [inline drivers/adc/adc_tla2021.c:166]: These are normally at the top of the function unless there's a good reason to declare them mid way, I'd suggest moving them. Same for `*data` below.

File: drivers/adc/adc_tla2021.c
Comment (fabiobaltieri) [inline drivers/adc/adc_tla2021.c:214]: I think `WRITE_BIT` is a bit uncommon if there's no condition, how about just `reg |= BIT(REG_CONFIG_MODE_pos)`?

File: drivers/adc/adc_tla2021.c
Comment (dp7hgh7) [reply drivers/adc/adc_tla2021.c:69]: Fair point. Done.

File: drivers/adc/adc_tla2021.c
Comment (dp7hgh7) [reply drivers/adc/adc_tla2021.c:116]: Fair point. Was a carry over from the old version.

File: drivers/adc/adc_tla2021.c
Comment (dp7hgh7) [reply drivers/adc/adc_tla2021.c:214]: What do you mean? `WRITE_BIT` comes from `util_macro.h`...

File: drivers/adc/adc_tla2021.c
Comment (dp7hgh7) [reply drivers/adc/adc_tla2021.c:166]: Don't really care ;-). Therefore: Done.

File: drivers/adc/adc_tla2021.c
Comment (dp7hgh7) [reply drivers/adc/adc_tla2021.c:74]: Can't be just `value` as it's the name of an argument. Changed it to `tmp`, though...

File: drivers/adc/adc_tla2021.c
Comment (fabiobaltieri) [reply drivers/adc/adc_tla2021.c:202]: I mean I find it odd to use `WRITE_BIT` with a fixed argument, like I normally think of this macro used to replace patterns like:

```
if (set) {
  val |= BIT(bit);
} else {
  val &= ~BIT(bit);
}
```

but you have no condition so why not just use `reg |= BIT(REG_CONFIG_MODE_pos) | BIT(REG_CONFIG_OS_pos)` in this case?

More of a preference, it just looks unusual to me.

Comment (dp7hgh7) [reply drivers/adc/adc_tla2021.c:202]: I see you point. But since it's static it doesn't add any complexity after compilation also manual bit operations would ad much more noise to the code.

Comment (fabiobaltieri) [reply drivers/adc/adc_tla2021.c:202]: Cool with that.

**Review by fabiobaltieri — DISMISSED** [review]

File: drivers/adc/Kconfig.tla2021
Comment (cfriedt) [inline drivers/adc/Kconfig.tla2021:1]: Not blocking, but typically we don't see email addresses in the copyright. That's up to you though.

**Review by cfriedt — DISMISSED** [review]
LGTM

File: drivers/adc/Kconfig.tla2021
Comment (dp7hgh7) [reply drivers/adc/Kconfig.tla2021:1]: true... but in order to not delay this any longer I would rather not change this.

**Review by cfriedt — APPROVED** [review]

**Review by fabiobaltieri — APPROVED** [review]

**Review by vaishnavachath — APPROVED** [review]

**Pull request merged**
